src/ZkFold/Base/Algebra/Polynomials/Multivariate.hs

```diff
@@ -28,35 +40,78 @@ import           ZkFold.Base.Algebra.Polynomials.Multivariate.Monomial
 
 -- | Monomial constructor
 monomial :: Monomial i j => Map i j -> M i j (Map i j)
 monomial = M . fromJust . toMonomial
 
 -- | Polynomial constructor
-polynomial :: Polynomial c i j => [(c, M i j (Map i j))] -> P c i j (Map i j) [(c, M i j (Map i j))]
-polynomial = sum . map (\m -> P [m]) . fromJust . toPolynomial
+polynomial ::
+    Polynomial c i j =>
+    [(c, M i j (Map i j))] ->
+    P c i j (Map i j) [(c, M i j (Map i j))]
+polynomial = sum . map (\m -> P [m]) . toPolynomial
 
 -- | @'var' i@ is a polynomial \(p(x) = x_i\)
 var :: Polynomial c i j => i -> P c i j (Map i j) [(c, M i j (Map i j))]
 var x = polynomial [(one, monomial (singleton x one))]
 
-evalMonomial :: forall i j m b . (FromMonomial i j m, MultiplicativeMonoid b, Exponent b j) => (i -> b) -> M i j m -> b
-evalMonomial f (M m) = product (map (\(i, j) -> f i ^ j) (toList $ fromMonomial @i @j m))
+evalMonomial :: forall i j m b .
+    FromMonomial i j m =>
+    MultiplicativeMonoid b =>
+    Exponent b j =>
+    (i -> b) -> M i j m -> b
+evalMonomial f (M m) = product
+    $ toList (fromMonomial @i @j m)
+        <&> (\(i, j) -> f i ^ j)
 
-evalPolynomial :: forall c i j m p b . (FromMonomial i j m, FromPolynomial c i j m p, Algebra c b, Exponent b j)
-    => (i -> b) -> P c i j m p -> b
-evalPolynomial f (P p) = sum $ map (\(c, m) -> scale c (evalMonomial f m)) (fromPolynomial @c @i @j @m @p p)
+evalPolynomial :: forall p c i j m b .
+    FromMonomial i j m =>
+    FromPolynomial c i j m p =>
+    Algebra c b =>
+    Exponent b j =>
+    (i -> b) -> P c i j m p -> b
+evalPolynomial f (P p) = sum
+    $ fromPolynomial @c @i @j @m @p p
+        <&> (\(c, m) -> scale c $ evalMonomial f m)
 
-variables :: forall c i j m p . (FromMonomial i j m, FromPolynomial c i j m p) => P c i j m p -> [i]
-variables (P p) = nubOrd $ concatMap (\(_, M m) -> keys (fromMonomial @i @j @m m)) $ fromPolynomial @c @i @j @m @p p
+variables :: forall c i j m p .
+    FromMonomial i j m =>
+    FromPolynomial c i j m p =>
+    P c i j m p -> [i]
+variables (P p) = nubOrd
+    . concatMap (\(_, M m) -> keys (fromMonomial @i @j @m m))
+    $ fromPolynomial @c @i @j @m @p p
 
-mapCoeffs :: forall c c' i j m p p' . (FromPolynomial c i j m p, ToPolynomial c' i j m p')
-    => (c -> c') -> P c i j m p -> P c' i j m p'
-mapCoeffs f (P p) = P . fromJust . toPolynomial $ map (first f) (fromPolynomial @c @i @j @m p)
+mapCoeffs :: forall c c' i j m p p' .
+    FromPolynomial c i j m p =>
+    ToPolynomial c' i j m p' =>
+    (c -> c') -> P c i j m p -> P c' i j m p'
+mapCoeffs f (P p) = P . toPolynomial
+    $ fromPolynomial @c @i @j @m p
+        <&> first f
+
+mapVarMonomial :: [Natural] -> Monomial' -> Monomial'
+mapVarMonomial vars (M as) = M $ mapKeys (mapVar vars) as
+
+mapVar :: [Natural] -> Natural -> Natural
+mapVar vars x = case x `elemIndex` vars of
+    Just i  -> i
+    Nothing -> error "mapVar: something went wrong"
+
+mapVarPolynomial :: [Natural] -> Polynomial' c -> Polynomial' c
```

Comment:
Maybe we can generalize this to `(i -> i') -> P c i j m p -> P c i' j m p'`?
(And once again, this can be replaced by a call to `evalPolynomial`, because this is like an "fmap for Polynomial monad")

Reply:
@TurtlePU We can but we don't use is now since in this PR `<To/From>Polynomial` classes are ditched already.